node/network/statement-distribution/src/vstaging/grid.rs

```diff
@@ -1518,20 +1503,533 @@ mod tests {
+
+		// Should start with no pending statements.
+		assert_eq!(tracker.pending_statements_for(validator_index, candidate_hash), None);
+		assert_eq!(tracker.all_pending_statements_for(validator_index), vec![]);
+
+		// Add the candidate as backed.
+		tracker.add_backed_candidate(
+			&session_topology,
+			candidate_hash,
+			group_index,
+			group_size,
+			local_knowledge.clone(),
+		);
+
+		// Import statement for originator.
+		let ack = tracker.import_manifest(
+			&session_topology,
+			&groups,
+			candidate_hash,
+			3,
+			ManifestSummary {
+				claimed_parent_hash: Hash::repeat_byte(0),
+				claimed_group_index: group_index,
+				seconded_in_group: bitvec::bitvec![u8, Lsb0; 0, 1, 0],
+				validated_in_group: bitvec::bitvec![u8, Lsb0; 1, 0, 1],
+			},
+			ManifestKind::Full,
+			validator_index,
+		);
+		tracker.manifest_sent_to(&groups, validator_index, candidate_hash, local_knowledge.clone());
+		let statement = CompactStatement::Seconded(candidate_hash);
+		tracker.learned_fresh_statement(&groups, &session_topology, validator_index, &statement);
+
+		// Import statement for counterparty.
+		let ack = tracker.import_manifest(
+			&session_topology,
+			&groups,
+			candidate_hash,
+			3,
+			ManifestSummary {
+				claimed_parent_hash: Hash::repeat_byte(0),
+				claimed_group_index: group_index,
+				seconded_in_group: bitvec::bitvec![u8, Lsb0; 0, 1, 0],
+				validated_in_group: bitvec::bitvec![u8, Lsb0; 1, 0, 1],
+			},
+			ManifestKind::Full,
+			counterparty,
+		);
+		tracker.manifest_sent_to(&groups, counterparty, candidate_hash, local_knowledge);
+		let statement = CompactStatement::Seconded(candidate_hash);
+		tracker.learned_fresh_statement(&groups, &session_topology, counterparty, &statement);
+
+		// There should be pending statements now.
+		let statements = StatementFilter {
+			seconded_in_group: bitvec::bitvec![u8, Lsb0; 1, 0, 0],
+			validated_in_group: bitvec::bitvec![u8, Lsb0; 0, 0, 0],
+		};
+		assert_eq!(
+			tracker.pending_statements_for(validator_index, candidate_hash),
+			Some(statements.clone())
+		);
+		assert_eq!(
+			tracker.all_pending_statements_for(validator_index),
+			vec![(ValidatorIndex(0), CompactStatement::Seconded(candidate_hash))]
+		);
+		assert_eq!(
+			tracker.pending_statements_for(counterparty, candidate_hash),
+			Some(statements.clone())
+		);
+		assert_eq!(
+			tracker.all_pending_statements_for(counterparty),
+			vec![(ValidatorIndex(0), CompactStatement::Seconded(candidate_hash))]
+		);
+
+		// Note sending or receiving of a direct statement.
+		tracker.sent_or_received_direct_statement(
+			&groups,
+			validator_index,
+			counterparty,
+			&statement,
+		);
+
+		// There should be no pending statements now (for the counterparty).
+		assert_eq!(tracker.all_pending_statements_for(counterparty), vec![]);
+		// TODO: There are still `pending_statements_for`, is this correct?
+		// assert_eq!(tracker.pending_statements_for(counterparty, candidate_hash), None);
```

Comment:
`all_pending_statements_for` is updated but not `pending_statements_for`, is this expected?

Reply:
It is wrong, but I think the reason is that `tracker.learned_fresh_statement` needs to be called before `sent_or_received_direct_statement`.